### PR TITLE
[JENKINS-46386] - Detach ServletContext logic to a pluggable Provider

### PR DESCRIPTION
**Pull request by oleg-nenashev**
It should prevent classloading issues on agents for sure. The implementation is a bit hackish, ideally providers should be somehow integrated into the `Jenkins` object singleton (though then we need to prevent variable usage before Jenkins' full initialization)

No autotests so far, because it requires specification of a custom `web.xml` for the WAR file. Will tests the case manually

See [JENKINS-46386](https://issues.jenkins-ci.org/browse/JENKINS-46386).

### Proposed changelog entries

* Bug: JENKINS-46386 - Prevent failure of custom agent types with  `javax.servlet.ServletContextListener`classloading issues. Now the class is not being propagated to agents
* ...

### Submitter checklist

- [x] JIRA issue is well described
- [x] Changelog entry appropriate for the audience affected by the change (users or developer, depending on the change). [Examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)
      * Use the `Internal: ` prefix if the change has no user-visible impact (API, test frameworks, etc.)
- [x] Appropriate autotests or explanation to why this change has no tests
- [x] For dependency updates: links to external changelogs and, if possible, full diffs

### Desired reviewers

@reviewbybees @KostyaSha

### CODE REVIEW

File: core/src/main/java/jenkins/util/SystemPropertiesProvider.java [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+public abstract class SystemPropertiesProvider {
+
+    /**
+     * Logger.
+     */
+    private static final Logger LOGGER = Logger.getLogger(SystemPropertiesProvider.class.getName());
+
+    //TODO: This should use the annotation indexer at least.
+    // Ideally it should be a part of the Jenkins singleton in order to play nicely with {@code JenkinsRule}.
+    private static final CopyOnWriteMap<String, SystemPropertiesProvider> providers = new CopyOnWriteMap.Hash<>();
```
Comment (jtnord) [inline core/src/main/java/jenkins/util/SystemPropertiesProvider.java]: the map symatics are not really unused - you could remove using the actual provider (object) to remove and use a standard CopyOnWrite collection.

Ontop of this you get different (random) iteration order (the CopyOnWriteMap makes not guarntees about ordering... so could get hard to reproduce bugs when you sometimes get a system property and other times get a ServletContext parameter...

**Review by jtnord — COMMENTED** [review]
left a comment about what looks like a bug inline.

ontop of this I wonder if anyone really uses this functionality.  Would a flat properties file in JENKINS_HOME (CWD of the agent) not be much better and would work on agents and masters alike?

(OT) The fact the jenkins changes behaviour based on unversioned environment variables also troubles me (I still see people starting services with `/etc/initi.d/servericec start`... which will not start with a clean environment but whatever the current user has polluted their env with to begin with

File: core/src/main/java/jenkins/util/SystemPropertiesProvider.java (hunk 1 above)
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/util/SystemPropertiesProvider.java]: Yeah, probably Map is not required. I just added it to support Removal, but there are more simple and efficient ways for sure

**Comment by oleg-nenashev** [comment]
> ontop of this I wonder if anyone really uses this functionality. Would a flat properties file in JENKINS_HOME (CWD of the agent) not be much better and would work on agents and masters alike?

Yes, I confirm I use these settings to customize the Jenkins behavior from my custom Jenkins WAR packaging. More in the private chat.

> The fact the jenkins changes behaviour based on unversioned environment variables also troubles me

I suggest filing a ticket

**Comment by oleg-nenashev** [comment]
I need to fix the test codebase. Yes, I will try compiling it on my machine next time...

**Comment by jglick** [comment]
The broader issue is that #3067 lacked a test because

* agents will only be unable to load `javax.servlet.*` under certain servlet containers or environments, IIUC
* Jenkins fails to block `RemoteClassLoader` from even trying to load such classes under _normal_ conditions
* thus, routine test suites which do simple stuff on agents pass despite the bug

This is long filed in JIRA and I would rather prioritize a fix of _that_ so we are not making speculative and untestable fixes.

**Comment by oleg-nenashev** [comment]
Yes, even on the wild the issue happens rarely. @KostyaSha got a reproducible scenario in YAD, but I was using the plugin easily for many containers without an issue. ATH does not catches it as well. @jglick are you fine with the current fix PoC in this PR?  

For sure, we need to make the classloading logic more efficient + maybe add some static analysis around the `OnMaster` interface. It's not the only place in the code we may get into such misbehavior, unfortunately.

File: core/src/main/java/jenkins/util/ServletContextSystemPropertiesProvider.java [hunk 2]
```diff
@@ -0,0 +1,74 @@
...
+     */
+    @CheckForNull
+    private static ServletContext theContext;
+
+    /**
+     * Called by the servlet container to initialize the {@link ServletContext}.
+     */
+    @Override
+    @SuppressFBWarnings(value = "ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD",
+            justification = "Currently Jenkins instance may have one ond only one context")
```
Comment (jglick) [inline core/src/main/java/jenkins/util/ServletContextSystemPropertiesProvider.java]: typo

File: core/src/main/java/jenkins/util/ServletContextSystemPropertiesProvider.java [hunk 3]
```diff
@@ -0,0 +1,74 @@
...
+@Restricted(NoExternalUse.class)
+public class ServletContextSystemPropertiesProvider extends SystemPropertiesProvider implements OnMaster, ServletContextListener {
+
+    // this class implements ServletContextListener and is declared in WEB-INF/web.xml
+
+    /**
+     * The ServletContext to get the "init" parameters from.
+     */
+    @CheckForNull
+    private static ServletContext theContext;
```
Comment (jglick) [inline core/src/main/java/jenkins/util/ServletContextSystemPropertiesProvider.java]: Why is this `static`?

File: core/src/main/java/jenkins/util/SystemPropertiesProvider.java [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+
+/**
+ * Custom Provider extension.
+ * @author Oleg Nenashev
+ * @see SystemProperties
+ * @see ServletContextSystemPropertiesProvider
+ */
+@Restricted(NoExternalUse.class)
+public abstract class SystemPropertiesProvider {
```
Comment (jglick) [inline core/src/main/java/jenkins/util/SystemPropertiesProvider.java]: This and `ServletContextSystemPropertiesProvider` could have just have been nested classes of `SystemProperties`, I think; would be more readable.

File: core/src/main/java/jenkins/util/SystemPropertiesProvider.java (hunk 1 above)
Comment (jglick) [reply core/src/main/java/jenkins/util/SystemPropertiesProvider.java]: Well, this is all restricted, and there is no removal, so KISS.

File: core/src/main/java/jenkins/util/SystemPropertiesProvider.java [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+        String providerId = providerClass.getName();
+        LOGGER.log(Level.INFO, "Registering new SystemPropertyProvider: {0}", providerId);
+        return providers.remove(providerId);
+    }
+
+    /**
+     * Gets all registered providers.
+     * @return Collection of registered providers.
+     */
+    public static Collection<SystemPropertiesProvider> all() {
```
Comment (jglick) [inline core/src/main/java/jenkins/util/SystemPropertiesProvider.java]: unused

File: core/src/main/java/jenkins/util/SystemPropertiesProvider.java [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+        return providers.put(providerId, provider);
+    }
+
+    /**
+     * Removes provider.
+     * @param providerClass Class of the provider to be removed.
+     * @return Removed provider (if any)
+     */
+    @CheckForNull
+    public static SystemPropertiesProvider removeProvider(@Nonnull Class<?> providerClass) {
```
Comment (jglick) [inline core/src/main/java/jenkins/util/SystemPropertiesProvider.java]: unused

File: test/src/test/java/jenkins/util/ServletContextSystemPropertiesProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,86 @@
...
+ */
+public class ServletContextSystemPropertiesProviderTest {
+
+    @Rule
+    public JenkinsRule j = new JenkinsRule();
+
+    @Before
+    public void setUp() {
+        // Does not matter what we call, the fields are static
+        new ServletContextSystemPropertiesProvider().contextInitialized(new ServletContextEvent(j.jenkins.servletContext));
```
Comment (jglick) [inline test/src/test/java/jenkins/util/ServletContextSystemPropertiesProviderTest.java]: Why would this not be called automatically by Jetty during `JenkinsRule` setup?

**Review by jglick — APPROVED** [review]
A bit overengineered, and still does nothing to block us from making the broader category of mistakes in the future (nor is there any test proving that the fix works), but seems to be a step in the right direction.

**Comment by jglick** [comment]
Except:

    Write to static field jenkins.util.ServletContextSystemPropertiesProvider.theContext from instance method jenkins.util.ServletContextSystemPropertiesProvider.contextDestroyed(ServletContextEvent) [jenkins.util.ServletContextSystemPropertiesProvider] At ServletContextSystemPropertiesProvider.java:[line 66] ST_WRITE_TO_STATIC_FROM_INSTANCE_METHOD

as I noted.

File: core/src/main/java/jenkins/util/SystemPropertiesProvider.java (hunk 5 above)
Comment (oleg-nenashev) [reply core/src/main/java/jenkins/util/SystemPropertiesProvider.java]: Restricted anyway

**Comment by oleg-nenashev** [comment]
Needs fix, yes. Due to whatever reason, FindBugs behaves differently in CI

**Comment by oleg-nenashev** [comment]
After #3362  this fix is not actual, but it still makes sense to revive it once we need better extensibility

**Review by stephenc — COMMENTED** [review]
Is this still relevant?

**Comment by batmat** [comment]
@oleg-nenashev also, it's conflicted. What do you think about closing it?

**Comment by oleg-nenashev** [comment]
Aside grumbling about priority switches and lots of almost-done work being abandoned, I am fine with that. Closing it

**Pull request closed without merging**